src/main/java/com/thomashan/collection/immutable/ImmutableListImpl.java

```diff
@@ -0,0 +1,183 @@
+package com.thomashan.collection.immutable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+class ImmutableListImpl<E> implements ImmutableList<E> {
+    private final transient List<E> list;
+
+    private ImmutableListImpl(List<E> list) {
+        this.list = list;
+    }
+
+    static <E> ImmutableListImpl<E> of(E... e) {
+        if (0 == e.length) {
+            return new ImmutableListImpl<>(Collections.unmodifiableList(Collections.emptyList()));
+        }
+
+        return new ImmutableListImpl(Collections.unmodifiableList(Arrays.asList(e)));
+    }
+
+    static <E> ImmutableListImpl<E> of(List<E> list) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/thomashan/coup-java/pullRequest?prid=4509131)